iothub/service/src/Amqp/AmqpConnectionHandler.cs

```diff
@@ -40,6 +40,9 @@ internal class AmqpConnectionHandler
         // The current delivery tag. Increments after each send operation to give a unique value.
         private int _sendingDeliveryTag;
 
+        // The lock that prevents simultaneous open/close, open/open, and close/close operations
+        private SemaphoreSlim _openCloseSemaphore = new SemaphoreSlim(1, 1);
```

Comment:
It's a bit regrettable that this is the sole reason why we need these amqp subclients to be disposable. Is there some reason why we don't tend to use the non-disposable [semaphore](https://docs.microsoft.com/en-us/dotnet/api/system.threading.semaphore?view=net-6.0) class for cases like these?

Reply:
The general guidance is generally that we should use SemaphoreSlim when we want to use local semaphores. For semaphores that are used across processes (system-wide) we'd want to use Semaphore. As for SemaphoreSlim being disposable and Semaphore not being disposable, that's a good point that I hadn't considered before. Something worth looking into for sure.

Reply:
Apparently ```Semaphore``` doesn't have cancellation token support, so I suppose we'll need to stick with ```SemaphoreSlim```

Reply:
Semaphore is IDisposable (via inheritance).